crates/energy/src/graph.rs

```diff
@@ -0,0 +1,259 @@
+        commands
+            .entity(entity)
+            .insert((EnergyReceiver, NearbyUnits::default()));
+    }
+}
+
+pub fn update_nearby_recv(
+    spacial_index_producer: SpatialQuery<Entity, With<EnergyProducer>>,
+    spacial_index_receiver: SpatialQuery<Entity, With<EnergyReceiver>>,
+    mut nearby_units: Query<(Entity, &mut NearbyUnits, &Transform), Changed<Transform>>,
+) {
+    nearby_units
+        .par_iter_mut()
+        .for_each_mut(|(entity, mut nearby_units, transform)| {
+            if let Some(last_pos) = nearby_units.last_pos {
+                if transform.translation.to_flat().distance_squared(last_pos) < 0.5 {
+                    return;
+                }
+            }
+            nearby_units.last_pos = Some(transform.translation.to_flat());
+
+            let aabb = &Aabb::new(
+                Point::from(transform.translation - Vec3::splat(MAX_DISTANCE)),
+                Point::from(transform.translation + Vec3::splat(MAX_DISTANCE)),
+            );
+
+            let producers = spacial_index_producer.query_aabb(aabb, Some(entity));
+
+            let receivers = spacial_index_receiver.query_aabb(aabb, Some(entity));
+
+            nearby_units.units.clear();
+
+            nearby_units
+                .units
+                .producers
+                .extend(producers.map(|entity| entity.into()));
+            nearby_units
+                .units
+                .receivers
+                .extend(receivers.map(|entity| entity.into()));
+        });
+}
+
+fn update_graph(
+    mut power_grid: ResMut<PowerGrid>,
+    nearby_units: Query<(Entity, &NearbyUnits), Changed<NearbyUnits>>,
+) {
+    for (entity, nearby_units) in nearby_units.iter() {
+        let mut edges_to_remove = HashSet::new();
+
+        edges_to_remove.extend(
+            &mut power_grid
+                .graph
+                .edges(entity)
+                .map(|edge| (edge.source(), edge.target())),
+        );
+
+        let mut edges_to_add = vec![];
+
+        for nearby_entity in nearby_units.units.both() {
+            edges_to_add.push((entity, nearby_entity.into()));
+        }
+
+        for edge in edges_to_add.iter() {
+            if !power_grid.graph.contains_edge(edge.0, edge.1) {
+                power_grid.graph.add_edge(edge.0, edge.1, 1.0);
+            }
+            if edges_to_remove.contains(edge) {
+                edges_to_remove.remove(edge);
+            }
+        }
+
+        for edge in edges_to_remove.iter() {
+            power_grid.graph.remove_edge(edge.0, edge.1);
+        }
+    }
+}
+
+fn remove_old_nodes(
+    mut power_grid: ResMut<PowerGrid>,
+    mut nearby_units: Query<&mut NearbyUnits>,
+    mut death_events: EventReader<DespawnedComponentsEvent<NearbyUnits>>,
+) {
+    for event in death_events.iter() {
+        power_grid.graph.remove_node(event.entity);
+
+        // Remove the entity from the nearby units of all nearby units
+        for mut nearby_units in nearby_units.iter_mut() {
```

Comment:
It is not necessary to update all of the. It would be enough (pseudocode):

```rust
let despawned_nearby = nearby_units.get(event.entity);
for producer in despawned_nearby.producers {
    nearby_units.get_mut(producer).receivers.remove(event.entity);
}
for receiver in despawned_nearby.receivers {
    nearby_units.get_mut(receiver).producers.remove(event.entity);
}
```